File: go/src/step1_read_print/step1_read_print.go
package main

import (
    "fmt"
    "strings"
)

import (
    "readline"
    . "types"
    "reader"
    "printer"
)

// read
func READ(str string) (MalType, error) {
    return reader.Read_str(str)
}

// eval
func EVAL(ast MalType, env string) (MalType, error) {
    return ast, nil
}

// print
func PRINT(exp MalType) (string, error) {
    return printer.Pr_str(exp, true), nil
}

// repl
func rep(str string) (MalType, error) {
    var exp MalType
    var res string
    var e error
    if exp, e = READ(str); e != nil { return nil, e }
    if exp, e = EVAL(exp, ""); e != nil { return nil, e }
    if res, e = PRINT(exp); e != nil { return nil, e }
    return res, nil
}

func main() {
    // repl loop
    for {
        text, err := readline.Readline("user> ")
        text = strings.TrimRight(text, "\n");
        if (err != nil) {
            return
        }
        var out MalType
        var e error
        if out, e = rep(text); e != nil {
            if e.Error() == "<empty line>" { continue }
            fmt.Printf("Error: %v\n", e)
            continue
        }
        fmt.Printf("%v\n", out)
    }
}

File: impls/deno/step1_read_print.ts
import * as Reader from "./reader.ts";
import * as Printer from "./printer.ts";
import { MalType } from "./types.ts";
import { readline } from "./readline.ts";


const read = (str: string): MalType => Reader.readStr(str);

const eval_ = (ast: MalType): MalType => ast;

const print = (exp: MalType): string => Printer.prStr(exp);

const rep = (str: string): string => print(eval_(read(str)));

while (true) {
  const value = readline("user> ");

  if (value === undefined) {
    break;
  } else if (value === "") {
    continue;
  }

  try {
    console.log(rep(value));
  } catch (e) {
    console.error(e.message);
  }
}
